doc/developers/minimal_reproducer.rst

```diff
@@ -0,0 +1,404 @@
+
+Provide a failing code example with minimal comments
+----------------------------------------------------
+
+Instructions may get lost in translation for non-native english speakers.
+Hence, it is better that all the necessary information can be read
+from the code itself. Besides, by this point you already provided a concise
+description in the **Describe the bug** section of the Issue template.
+
+**Improved example**
+
+.. code-block:: python
+
+    import pandas as pd
+    df = pd.read_csv('my_data.csv')
+    X = df[["feature_name"]]
+    y = df["target"]
+
+    from sklearn.model_selection import train_test_split
+    X_train, X_test, y_train, y_test = train_test_split(
+        X, y, test_size=0.33, random_state=42)
+
+    from sklearn.ensemble import GradientBoostingRegressor
+
+    gbdt = GradientBoostingRegressor(random_state=0)
+    gbdt.fit(X_train, y_train) # compiles
+
+    gbdt = GradientBoostingRegressor(random_state=0, n_iter_no_change=5)
+    gbdt.fit(X_train, y_train) # raises warning
+
+
+Boil down your script to something as small as possible
+-------------------------------------------------------
+
+You have to ask yourself which lines of code are relevant and which are not for
+raising the bug. Erasing unnecessary lines of code will help you and other
+contributors narrow down the bug. In this example the warning has nothing to
+do with the `train_test_split`, nor the `random_state`.
+
+**Improved example**
+
+.. code-block:: python
+
+    import pandas as pd
+    df = pd.read_csv('my_data.csv')
+    X = df[["feature_name"]]
+    y = df["target"]
+
+    from sklearn.ensemble import GradientBoostingRegressor
+
+    gbdt = GradientBoostingRegressor()
+    gbdt.fit(X, y) # compiles
+
+    gbdt = GradientBoostingRegressor(n_iter_no_change=5)
+    gbdt.fit(X, y) # raises warning
+
+
+**DO NOT** report your data unless it is extremely necessary
+------------------------------------------------------------
+
+The idea is to make the code as self-contained as possible. For doing so, you
+can use a :ref:`synth_data`. It can be generated using numpy, pandas or the
+:mod:`sklearn.datasets` module. Most of the times the bug is not related to a
+particular structure of your data. Even if it is, try to find an available
+dataset that has similar characteristics to yours and that reproduces the
+problem. In this particular case, we are interested in data that has labeled
+feature names.
+
+**Improved example**
+
+.. code-block:: python
+
+    import pandas as pd
+    df = pd.DataFrame({"feature_name": [-12.32, 1.43, 30.01, 22.17], "target": [72, 55, 32, 43]})
+    X = df[["feature_name"]]
+    y = df["target"]
+
+    from sklearn.ensemble import GradientBoostingRegressor
+
+    gbdt = GradientBoostingRegressor()
+    gbdt.fit(X, y) # compiles
+
+    gbdt = GradientBoostingRegressor(n_iter_no_change=5)
+    gbdt.fit(X, y) # raises warning
+
+Improve your formatting as much as possible
+-------------------------------------------
```

Comment:
I am not so sure this step is necessary.

While following code conventions is important when contributing a PR, it's not as important when crafting a minimal reproducer. I would therefore shorten this tutorial by removing this step and not mentioning PEP8 and other formatting concerns.

Instead we could directly improve the formatting of the previous step to directly show a good example of well formatted minimal reproducer implicitly.

Reply:
I tried to keep the guide relatively general for concordance with the first paragraph:

> Whether submitting a bug report, designing a suite of tests, or simply posting a
question in the discussions, being able to craft minimal, reproducible examples
(or minimal, workable examples) is the key to communicating effectively and
efficiently with the community.

I don't have strong feelings about keeping this step, but if you think it is making this guide too long, maybe we can send this part to the end of the page as an "extra"/"nice-to-have", what do you think?

Reply:
I think it's better to remove the section on formatting and use a better code style from the beginning (show by example) and focus the content of this guide on the most important points (simplifying the code and test data, including the imports).

Let's try not be too nitpicky for bug reports.